### ISSUE TITLE
Show recent action summary at top of pages

### ISSUE THREAD

**Issue by ahelsing** [GENI-NSF/geni-portal#120]
After you take an action, you get no feedback that it worked.
Showing a message at the top would help.
EG 'Keys uploaded' or 'SSH keys generated' or 'Slivers created' or 'Slivers deleted' or something.
Is this only your own actions? Or other people acting on the same sliver or project? Is this just on the slice, project and profile pages?

We should talk to Joanne about how to make this look decent.

 _Imported from trac ticket [#120](http://trac.gpolab.bbn.com/proto-ch/ticket/120),  created by ahelsing on 06-07-2012 at 16:34, last modified: 06-12-2012 at 12:53_

**Comment by ahelsing** [comment]
After you handle a project request is another place like this.


_Trac comment by ahelsing on 06-07-2012 at 16:39_

**Comment by ahelsing** [comment]
This could be the last N logs in any time frame.

Or logs in the last X days, max of N logs.

Or only show what was just done by this user (previous page passes a message along).


_Trac comment by ahelsing on 06-08-2012 at 12:06_

**Comment by ahelsing** [comment]
Tom says:
Use cookies.
Pages that take actions set a cookie.
Pages that follow action pages (slices, projects, slice, project, profile, home) include a tool-lastaction which checks that cookie, prints the last action at the top, and clears the cookie.

The result is that the single last action is printed on the very next page, but never again.

See the edit-project / do-edit-project pages which may need to use this same mechanism.


_Trac comment by ahelsing on 06-11-2012 at 14:16_

**Comment by tcmitchell** [comment]
I may have been thinking "session" when I said "cookie". In other words, maybe we can tuck the string into the session along with the cached info that we've already got in the session. That piggybacks on an existing capability rather than trying to use a new capability (cookies). Consider both cookies and session as possibilities for implementing this.

_Trac comment by tmitchel (github user: tcmitchell) on 06-11-2012 at 15:27_

**Comment by ahelsing** [comment]
I've been trying the session thing.

Some notes:

 - Calling session_start() a 2nd time results in warnings
 - So far, I haven't managed to get the string I stuff into the session back out again.
 - see ticket #118
 - Basic plan is to add a tool-showmessage.php after tool-breadcrumbs.php on home, profile, project, slice, maybe projects and slices, of roughly this:
{{{
if (isset($_SESSION['lastmessage'])) {
  print $_SESSION['lastmessage'] . "<br/>\n";
  unset($_SESSION['lastmessage']);
}
 - Then places that do a redirect or download just set that field.

Places with edits include:
certificate.php:
`$_SESSION['lastmessage'] = "Downloaded certificate to $filename";`
createslice.php:
`$_SESSION['lastmessage'] = "Created slice $slice_name";`
deletesshkey.php:
`$_SESSION['lastmessage'] = "Deleted SSH Key $key_id: $result";`
do-disable-project.php:
`$_SESSION['lastmessage'] = "Asked to disable project $project_name - NOT IMPLEMENTED";`
do-disable-slice
do-edit-project
downloadkeycert
generatesshkey
handle-project-request
uploadsshkey



_Trac comment by ahelsing on 06-11-2012 at 16:59_

**Comment by ahelsing** [comment]
Keeping notes here:
user.php and sr_client.php reset $_SESSION to array() if they call session_start(). That seems wrong to me, but maybe I'm wrong. That would seem like it loses the state.

But if I take that out, I get an error in the log:
certificate.php does:
```

require_once("user.php");
error_log("cert back from require user");
$user = geni_loadUser();
error_log("cert did loadUser");
if (isset($user) || is_null($user) || ! $user->isActive()) {
  relative_redirect('home.php');
}
error_log("did check for isACtive")
```

log shows:
```

[Tue Jun 12 10:58:43 2012] [error] [client 128.89.81.32] cert back from require user
[Tue Jun 12 10:58:43 2012] [error] [client 128.89.81.32] cert did loadUser
[Tue Jun 12 10:58:43 2012] [error] [client 128.89.81.32] PHP Fatal error:  main(): The script tried to execute a method or access a property of an incomplete object. Please ensure that the class definition "GeniUser" of the object you are trying to operate on was loaded _before_ unserialize() gets called or provide a __autoload() function to load the class definition  in /var/www/secure/certificate.php on line 49
```

Diagnosis: The 'status' field on the user does not exist. Following the code, I suspect a bug in geni_loadUser_cache()

Putting in more printouts, I see:
```
[Tue Jun 12 11:06:00 2012] [error] [client 128.89.81.32] pulled user from user_cache_eppn from session
[Tue Jun 12 11:06:00 2012] [error] [client 128.89.81.32] CACHE ACCT= EPPN=ahelsing@gpolab.bbn.com USER=__PHP_Incomplete_Class Object\n(\n    [__PHP_Incomplete_Class_Name] => GeniUser\n    [identity_id] => 1\n    
```

So geni_loadUser_cache is finding the user not by account_id, but by eppn. But the thing in there is clearly labeled an incomplete_class object. Which is what PHP complains about later.

As usual, Stackoverflow has the answer I want:
```
by the time the session_start() call is made, the class declaration/definition has already been encountered by PHP. otherwise it would not be able to deserialize the object from the session store.
```

In other words, we have to load the GeniUser class definition before calling session_start(), or else it cannot load from the cache.

That fixes the problem.

Note that Stackoverflow commenters argue strongly against using sessions for lots of stuff, due to the cost of deserializing objects.

_Trac comment by ahelsing on 06-12-2012 at 11:18_

**Comment by ahelsing** [comment]
OK, fixed I think.

user.php is the canonical place to do session_start. sr_client does it too, but cannot use user.php (wrong dir, wrong mix of server & client side code).

user.php must define Geniuser before it deserializes the user

session_start resumes the session. Do not clear $_SESSION: let session_start either initialize it empty or fill it from the cache.

Once that is all in, using the session here works.
I added a new 'lastmessage' slot in a bunch of places - basically where the page does a download file or a relative_redirect



_Trac comment by ahelsing on 06-12-2012 at 12:53_
